### PR TITLE
Implementation of storage pools for pylxd

### LINKED ISSUE #274: Storage pool methods need adding and storage-pools should only work with the 'storage' api_extension
Firstly, the storage-pools functionality should only be available if the 'storage' api_extension is available on the connected LXD server.

Secondly, there are a whole bunch of methods that are available that can be implemented:

* PUT /1.0/storage-pools/<name>  -- replace a storage pool configuration
* PATCH /1.0/storage-pools/<name> -- update bits of a storage pool configuration.
* DELETE /1.0/storage-pools/<name> -- delete a storage pool.
* GET /1.0/storage-pools/<name>/resources - return a list of resources available to a storage pool.
* GET /1.0/storage-pools/<name>volumes - return a list of storage pool volumes.
* POST /1.0/storage-pool/<name>/volumes - create a new storage pool volume
* POST /1.0/storage-pools/<pool>/volumes/<type>/<name> - rename a storage volume on a pool
* GET /1.0/storage-pools/<pool>/volumes/<type>/<name> - get information about a specific volume
* PUT /1.0/storage-pools/<pool>/volumes/<type>/<name> - replace a storage volume config
* PATCH /1.0/storage-pools/<pool>/volumes/<type>/<name> - update a bit of a storage volume config
* DELETE /1.0/storage-pools/<pool>/volumes/<type>/<name> - delete a storage volume.

### PR DESCRIPTION
**Pull request by ajkavanagh**
This fully completes the 'storage' and 'resources' api_extensions of lxd
and also introduces a patch method into the Client class and extends the
put and patch methods to the storage volume (as they'll be needed to do
sane modifications).

This PR has:
  * Code changes
  * Documentation changes
  * Unit tests
  * Integration tests for 16.04 and 18.04.

Fixes #274

### CODE REVIEW

File: doc/source/storage-pools.rst [hunk 1]
```diff
@@ -2,16 +2,41 @@ Storage Pools
 =============
 
 LXD supports creating and managing storage pools and storage volumes. General
-keys are top-level. Driver specific keys are namespaced by driver name. Volume
+keys are top-level. Driver specific keys are name spaced by driver name. Volume
```
Comment (ChrisMacNaughton) [inline doc/source/storage-pools.rst]: Is this necessary? `namespace` _is_ a valid word, afterall

File: doc/source/storage-pools.rst [hunk 2]
```diff
@@ -21,25 +46,103 @@ Storage-pools can be queried through the following client manager methods:
   - `create()` - Create a storage-pool.  **Note the config in the create class
     method is the WHOLE json object described as `input` in the API docs.**
     e.g. the 'config' key in the API docs would actually be `config.config` as
-    passed to this method.
+    passed to this methods.
```
Comment (ChrisMacNaughton) [inline doc/source/storage-pools.rst]: invalid plural?

File: pylxd/exceptions.py [hunk 3]
```diff
@@ -37,6 +40,17 @@ class LXDAPIExtensionNotAvailable(Exception):
     """An exception raised when requested LXD API Extension is not present
     on current host."""
 
+    def __init__(self, name, *args, **kwargs):
+        """Custom exception handling of the message is to convert the name in
```
Comment (ChrisMacNaughton) [inline pylxd/exceptions.py]: "in to" should read "into"

Comment (ajkavanagh) [reply pylxd/exceptions.py]: Good catch; I'm always getting 'in to' and 'into' confused.  This was useful to me: https://blog.oxforddictionaries.com/2015/08/10/in-to-versus-into/

**Pull request merged**